openvdb_points/tools/AttributeGroup.h

```diff
@@ -0,0 +1,227 @@
+//
+// *     Neither the name of Double Negative Visual Effects nor the names
+// of its contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// IN NO EVENT SHALL THE COPYRIGHT HOLDERS' AND CONTRIBUTORS' AGGREGATE
+// LIABILITY FOR ALL CLAIMS REGARDLESS OF THEIR BASIS EXCEED US$250.00.
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @file AttributeGroup.h
+///
+/// @authors Dan Bailey
+///
+/// @brief  Attribute Group access and filtering for iteration.
+///
+
+
+#ifndef OPENVDB_TOOLS_ATTRIBUTE_GROUP_HAS_BEEN_INCLUDED
+#define OPENVDB_TOOLS_ATTRIBUTE_GROUP_HAS_BEEN_INCLUDED
+
+#include <openvdb/openvdb.h>
+#include <openvdb_points/tools/AttributeArray.h>
+
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+
+typedef uint8_t GroupType;
+
+
+////////////////////////////////////////
+
+
+class GroupAttributeArray : public TypedAttributeArray<GroupType, NullAttributeCodec<GroupType> >
+{
+public:
+    /// Default constructor, always constructs a uniform attribute.
+    explicit GroupAttributeArray(size_t n = 1,
+        const ValueType& uniformValue = zeroVal<ValueType>())
+        : TypedAttributeArray<GroupType, NullAttributeCodec<GroupType> >(n, uniformValue) { }
+    /// Deep copy constructor (optionally decompress during copy).
+    GroupAttributeArray(const TypedAttributeArray& array, const bool decompress = false)
+        : TypedAttributeArray<GroupType, NullAttributeCodec<GroupType> >(array, decompress) { }
+
+    /// Cast an AttributeArray to GroupAttributeArray
+    static GroupAttributeArray& cast(AttributeArray& attributeArray);
+
+    /// Cast an AttributeArray to GroupAttributeArray
+    static const GroupAttributeArray& cast(const AttributeArray& attributeArray);
+
+    /// Return @c true if the AttributeArray provided is a group
+    static bool isGroup(const AttributeArray& attributeArray);
+
+    /// @brief Specify whether this attribute is for tracking group membership
+    /// @note  Attributes are not group attributes by default.
+    void setGroup(bool state);
```

Comment:
It may be worth simplifying the API a bit here, by removing the 'setGroup' method altogether, and instead having the GroupAttributeArray set its Group flag on construction. This would save the caller from having to explicitly call setGroup(true) after constructing a GroupAttributeArray. Whilst I've seen it in the unit tests, in practise, I would expect the use of GroupAttributeArray without the Group flag set to be pretty rare...